src/message/messagesReducer.js

```diff
@@ -17,15 +17,55 @@ import {
   EVENT_UPDATE_MESSAGE,
 } from '../actionConstants';
 import { NULL_ARRAY } from '../nullObjects';
+import { getNarrowsForMessage, keyFromNarrow } from '../utils/narrow';
 
 const initialState: MessagesState = Immutable.Map([]);
 
 const eventNewMessage = (state, action) => {
-  // TODO: Optimize -- Only update if the new message belongs to at least
-  // one narrow that is caught up.
+  const { message, caughtUp } = action;
+  const { flags } = message;
+
+  if (!flags) {
+    throw new Error('EVENT_NEW_MESSAGE message missing flags');
+  }
+
+  // Don't add a message that's already been added. It's probably
+  // very rare for a message to have already been added when we
+  // get an EVENT_NEW_MESSAGE, and perhaps impossible. (TODO:
+  // investigate?)
   if (state.get(action.message.id)) {
     return state;
   }
+
+  const narrowsForMessage = getNarrowsForMessage(message, action.ownUser, flags);
+  // `narrowsForMessage` will always have at least length 1 because of
+  // the home narrow.
```

Comment:
Ah, I suppose that could be removed. I found myself wondering what [`Array.prototype.some`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/some) does with an empty array (turns out it returns `false` for any condition), and I thought it might be helpful to the next reader if they knew they didn't have to wonder about it.

Reply:
Got it. I think it's best to see that behavior as the only way that `some` could reasonably work:
* Its job is to return true just if there's some element that the condition is true for.
* So if there are no elements at all, then there are certainly none that the condition is true for, and the return value should be false.

Or from another angle:
* Taking away elements from an array should only ever turn the return value of `some` from true to false, never from false to true.
* So if there's any array where `some` returns false for a given condition, then taking away all its elements should leave it still false.
* Moreover there was just one element the condition was true for and you take that one away, the return value should become false.
* So if you have an array where `some` returns true, and you take a bunch of elements away and you're down to just one left, then even if the return value is true at that point, it should be false after you take that last element away.

Reply:
Oh OK, makes sense! Thanks.

Reply:
> I think it's best to see that behavior as the only way that `some` could reasonably work:
> 
> * Its job is to return true just if there's some element that the condition is true for.
> * So if there are no elements at all, then there are certainly none that the condition is true for, and the return value should be false.

This is also basically exactly what we want to know in this piece of code, too:
* We want to add the message just if there's some narrow it belongs in where we're caught up.
* So we get an array of narrows this message belongs in, and we ask if any of them is a narrow where we're caught up.

If for some reason there were no narrows this message belongs in, then the answer would be no, there's no narrow this message belongs in where we're caught up. Which is exactly what `Array#some` gives us.